packages/liferay-npm-scripts/src/utils/deep-merge.js

```diff
@@ -35,14 +35,125 @@ function combineMerge(target, source, options) {
+	} catch (error) {
+		// Could be a circular reference, but we're unlikely to ever get here
+		// because the deepmerge package itself will die first.
+		return `[unstringifiable item: ${error}]`;
+	}
+}
+
+function getBabelName(item) {
+	if (typeof item === 'string') {
+		return item;
+	} else if (Array.isArray(item) && typeof item[0] === 'string') {
+		return item[0];
+	} else {
+		throw new Error(
+			`getBabelName(): malformed item ${getItemDescription(item)}`
+		);
+	}
+}
+
+function getBabelOptions(item) {
+	if (typeof item === 'string') {
+		return null;
+	} else if (Array.isArray(item)) {
+		const options = item[1];
+		return options &&
+			typeof options === 'object' &&
+			Object.keys(options).length
+			? options
+			: null;
+	} else {
+		// We never expect to get here, but just in case...
+		throw new Error('getBabelOptions(): incompatible item type');
+	}
+}
+
+/**
+ * Custom merge that knows how to merge "plugins" and "presets".
+ */
+function babelMerge(key) {
+	if (key === 'plugins' || key === 'presets') {
+		return function(target, source, options) {
+			// Create a mutable copy of `source`.
+			const pending = source.slice();
+
+			const result = target.map(targetItem => {
+				const targetName = getBabelName(targetItem);
+				const sourceIndex = pending.findIndex(sourceItem => {
+					const sourceName = getBabelName(sourceItem);
+					return sourceName === targetName;
+				});
+				if (sourceIndex !== -1) {
+					const [sourceItem] = pending.splice(sourceIndex, 1);
+					const mergedOptions = merge.all(
+						[
+							getBabelOptions(targetItem),
+							getBabelOptions(sourceItem)
+						].filter(Boolean),
+						options
+					);
+
+					return Object.keys(mergedOptions).length
+						? [targetName, mergedOptions]
+						: targetName;
+				} else {
+					const targetOptions = getBabelOptions(targetItem);
+					return targetOptions ? targetItem : targetName;
+				}
+			});
+
+			return result.concat(
+				pending.map(item => {
+					const itemName = getBabelName(item);
+					const itemOptions = getBabelOptions(item);
+					return itemOptions ? [itemName, itemOptions] : itemName;
+				})
+			);
+		};
+	} else {
+		return combineMerge;
+	}
+}
+
+const MODE = Object.freeze({
+	DEFAULT: 0,
+	OVERWRITE_ARRAYS: 1,
+	BABEL: 2
+});
+
 /**
  * Helper to get merge two json objects
- * @param {Object} defaultConfig Config file
- * @param {Object} customConfig Config file
+ * @param {Array} items An array of config objects
+ * @param {Number} mode Merge strategy for combining values.
  * @returns {Object}
  */
-module.exports = function(items, overwrite) {
-	return merge.all(items, {
-		arrayMerge: overwrite ? overwriteMerge : combineMerge
-	});
-};
+function deepMerge(items, mode = MODE.DEFAULT) {
+	switch (mode) {
+		case MODE.DEFAULT:
+			return merge.all(items, {
+				arrayMerge: combineMerge
+			});
+
+		case MODE.OVERWRITE_ARRAYS:
+			return merge.all(items, {
+				arrayMerge: overwriteMerge
+			});
+
+		case MODE.BABEL:
```

Comment:
Seems a little weird at first to make an exception for just babel here, could just be me though. However, the API is clear and makes sense with the whole `customMerge` option with `deepmerge`. Good idea with the `MODE` API.

Reply:
If you can think of anything more general than this we should implement it. Might be challenging though to find something that knows how to merge arrays where elements like `'foo'`  and `['foo', {option: true}]` actually refer to the same "thing".